yarn-project/simulator/src/public/public_tx_simulator.ts

```diff
@@ -0,0 +1,263 @@
+    ) as ProcessedPhase[];
+
+    const _endStateReference = await this.db.getStateReference();
+    const transactionFee = context.getTransactionFee();
+
+    const tailKernelOutput = await this.publicKernelTailSimulator.simulate(context.latestPublicKernelOutput);
+
+    context.avmProvingRequest!.inputs.output = generateAvmCircuitPublicInputs(
+      tx,
+      tailKernelOutput,
+      context.getGasUsedForFee(),
+      transactionFee,
+    );
+
+    const gasUsed = {
+      totalGas: context.getActualGasUsed(),
+      teardownGas: context.teardownGasUsed,
+    };
+    return {
+      avmProvingRequest: context.avmProvingRequest!,
+      gasUsed,
+      revertCode: context.revertCode,
+      revertReason: context.revertReason,
+      processedPhases: processedPhases,
+    };
+  }
+
+  private async processSetupPhase(context: PublicTxContext): Promise<ProcessedPhase | undefined> {
+    // Start in phase TxExecutionPhase.SETUP;
+    if (context.hasPhase()) {
+      return await this.processPhase(context);
+    }
+  }
+
+  private async processAppLogicPhase(context: PublicTxContext): Promise<ProcessedPhase | undefined> {
+    context.progressToNextPhase(); // to app logic
+    if (context.hasPhase()) {
+      // Fork the state manager so that we can rollback state if app logic or teardown reverts.
+      // Don't need to fork for setup since it's non-revertible (if setup fails, transaction is thrown out).
+      context.state.fork();
+
+      const result = await this.processPhase(context);
+
+      if (result.reverted) {
+        // Drop the currently active forked state manager and rollback to end of setup.
+        // Fork again for teardown so that if teardown fails we can again rollback to end of setup.
+        context.state.discardForkedState();
+      } else {
+        if (!context.hasPhase(TxExecutionPhase.TEARDOWN)) {
+          // Nothing to do after this (no teardown), so merge state in now instead of letting teardown handle it.
+          context.state.mergeForkedState();
+        }
+      }
+
+      return result;
+    }
+  }
+
+  private async processTeardownPhase(context: PublicTxContext): Promise<ProcessedPhase | undefined> {
+    context.progressToNextPhase(); // to teardown
+    if (context.hasPhase()) {
+      if (!context.state.isForked()) {
+        // if state isn't forked (app logic was empty or reverted), fork now
+        // so we can rollback to the end of setup on teardown revert
+        context.state.fork();
+      }
+
+      const result = await this.processPhase(context);
+
+      if (result.reverted) {
+        // Drop the currently active forked state manager and rollback to end of setup.
+        context.state.discardForkedState();
+      } else {
+        context.state.mergeForkedState();
+      }
+
+      return result;
+    }
+  }
+
+  private async processPhase(context: PublicTxContext): Promise<ProcessedPhase> {
+    const tx = context.tx;
+    const callRequests = context.getCallRequestsForCurrentPhase();
```

Comment:
I'll take a look again but it feels like this code might be better and more explicit if the context wasn't a kind of state machine. Ideally you'd write here context.getCallRequestForPhase(phase).

You could try to make the PublicTxSimulator know about the transition of the phases, and the context just hold data. Now both know about the transitions.